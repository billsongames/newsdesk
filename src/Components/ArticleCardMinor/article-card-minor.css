.article-card-minor {
  margin: 10px;
  padding: 10px;
  background-color: var(--background);
  border: 2px solid var(--text);
  border-radius: 8px;
}

.article-card-minor__title {
  margin: 0px 0px 10px 0px;
  font-size: 18px;
  font-weight: bold;
  color: var(--text);
  opacity: var(--opacityHigh);
}

.article-card-minor__title-url {
  color: var(--text);
}

.article-card-minor__display {
  display: flex;
  flex-flow: row nowrap;
  justify-content: flex-start;
}

.article-card-minor__image > img {
  width: 160px;
  height: 120px;
  object-fit: cover;
  color: var(--text);
}

.article-card-minor__text {
  display: flex;
  flex-flow: column;
  color: var(--text);
}

.article-card-minor__description {
  font-weight: bold;
  margin: 0px 10px 10px 10px;
  font-size: 16px;
  color: var(--text);
  opacity: var(--opacityMedium);
}

.article-card-minor__content {
  margin: 10px;
  opacity: var(--opacityMedium);
}

.article-card-minor__url {
  margin: 10px;
  padding: 2px;
  font-size: 14px;
  color: var(--text);
  opacity: var(--opacityHigh);
}

.article-card-minor__source {
  padding: 2px;
  font-weight: bold;
  font-size: 14px;
  color: var(--text);
  opacity: var(--opacityMedium);
}

.article-card-minor__time {
  font-weight: bold;
  font-size: 14px;
  color: var(--text);
  opacity: var(--opacityMedium);
}

.article-buttons{
  display: flex;
  justify-content: right;
  height: 34px;
}

.social-media {
  font-size: 12px;
  font-weight: bold;
}

.social-media > *{
  margin: 0px 6px 0px 6px;
  transition-duration: 0.2s;
}

.social-media > *:hover{
  scale: 1.2
}


.article-save-button-login-prompt {
  border: none;
  background-color: var(--background);
  color: var(--text);
  margin: 0px 10px 0px 0px;
  width: 250px;
  height: 32px;
  font-weight: bold;
  transition-duration: 0.2s;
}


.article-save-button {
  background-color: var(--saveButtonColor);
  color: #FFFFFF;
  margin: 0px 10px 0px 0px;
  width: 120px;
  height: 32px;
  font-weight: bold;
  border: 2px solid var(--greenButtonBorder);;
  border-radius: 8px;
  transition-duration: 0.2s;
  opacity: var(opacityHigh);
}

.article-save-button:hover{
  scale: 1.1
}

.article-saved-button {
  background-color: var(--savedArticleButtonColor);
  color: #FFFFFF;
  margin: 0px 10px 0px 0px;
  width: 120px;
  height: 32px;
  font-weight: bold;
  border: solid 4px var(--savedArticleButtonColor);
  border-radius: 8px;
  transition-duration: 0.2s;
}






