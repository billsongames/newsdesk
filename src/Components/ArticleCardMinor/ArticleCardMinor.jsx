import React, { useEffect, useState } from "react";
import PropTypes from "prop-types"
import { supabase } from '../../api/api';
import { tester } from "../../api/api";
import { 
  FacebookShareButton, 
  FacebookIcon, 
  TwitterShareButton, 
  TwitterIcon, 
  WhatsappShareButton, WhatsappIcon,
  RedditShareButton, RedditIcon,
  EmailShareButton, 
  EmailIcon } from "react-share";

import "./article-card-minor.css"

function ArticleCardMinor( {title, description, content, image, url, source, publishedAt, userID} ) {

  const articleJSON = {"title": `${title}`, "image_url": `${image}`, "article_url": `${url}`, "publishedAt": `${publishedAt}`}
  const [alert,setAlert] = useState({message: ""})
  const [savedArticles, setSavedArticles] = useState([])
  const [alreadySaved, setAlreadySaved] = useState(0)
  
  useEffect(() => {
    setTimeout(() => {
      async function getUserSavedArticles() {
        if (userID) {
          const {data, error} = await supabase
            .from('users')
            .select('saved_articles')
            .eq('user_id', `${userID}`)
      
            if (error) {
              console.log("error", error)
              setAlert({message: "Error retrieving saved articles"})
            } else {
              setSavedArticles(data[0].saved_articles)
              setAlert({message: ""})
            }
          }
      }
      getUserSavedArticles()

      setAlreadySaved(savedArticles.findIndex(article => 
        article.title === title))

    }, 500)

  }, [userID, savedArticles, title])

  const updateSavedArticles = async() => {
    const newArticle=[articleJSON, ...savedArticles]
    const {data, error} = await supabase

      .from('users')
      .update({
        saved_articles: newArticle
      })
      .eq('user_id', userID)

      if (error) {
        console.log("error", error)
      } 
    }


  let interval = (Date.now() - Date.parse(publishedAt)) / 1000 / 60 / 60;
  let timeSincePublication=""

//MINUTES
  if (interval > 0 && interval < 1) {
    interval = Math.ceil(interval * 60)
    timeSincePublication=`${interval} minutes`

// HOUR
  if (interval >= 1 && interval < 2) {
    interval = Math.floor(interval)
    timeSincePublication=`${interval} hour`
  } 
  } 
// HOURS
  if (interval >= 2 && interval < 24) {
    interval = Math.floor(interval)
    timeSincePublication=`${interval} hours`
  } 

// DAY  
  if (interval >= 24 && interval <48) {
    interval=Math.floor(interval/24)
    timeSincePublication=`${interval} day`
  }

  // DAYS  
  if (interval >= 48) {
    interval=Math.floor(interval/24)
    timeSincePublication=`${interval} days`
  }

  return(
    <div className="article-card-minor">
      
      <div className="article-card-minor__title">
        {title}
      </div>

      <div className="article-card-minor__display">
        <div className="article-card-minor__image">
          <img className="image"src={image} alt={description}></img>        
        </div>
        <div className="article-card-minor__text">
          <div className="article-card-minor__description">
            {description}
          </div>
          <div className="article-card-minor__content">
            {content} <a href={url} target="blank" className="article-card-minor__url">Read full article</a> 
          </div>  
          
        </div>

      </div>

      <div className="article-card-minor__source">
        Source: {source}
      </div>
      
      <div className="article-card-minor__time">
        {timeSincePublication} ago
      </div>

      <div className="article-buttons">
        <div>
          <>
          {userID && alreadySaved >=0
          ?
          <button className="article-saved-button" onClick={updateSavedArticles} disabled>Article saved</button>
          : <></>
          }

          {userID && alreadySaved === -1
          ?
          <button className="article-save-button" onClick={updateSavedArticles}>Save</button>
          : <></>
          }
          </>
        </div>

        <div className="social-media">
          <>
          {userID
          ?
          <>
          <FacebookShareButton
            url={url}
            quote={title}
            hashtag="#news"
            >
            <FacebookIcon size={30} round />
          </FacebookShareButton>

          <TwitterShareButton
            url={url}
            quote={title}
            hashtag="#news"
            >
            <TwitterIcon size={30} round />
          </TwitterShareButton>

          <WhatsappShareButton
            url={url}
            quote={title}
            >
            <WhatsappIcon size={30} round/>
          </WhatsappShareButton>

          <RedditShareButton
            url={url}
            title={title}
            >           
            <RedditIcon size={30} round />
          </RedditShareButton>  

          <EmailShareButton
            subject={title}
            body={`${description}`}
            >
            <EmailIcon size={30} round/>
          </EmailShareButton>
          </>
          : <><button className="article-save-button-login-prompt" onClick={updateSavedArticles} disabled>Log in to share and save articles</button></>
          }
          </>
        </div>
      </div>  

  </div>
  )
      
}
  


export default ArticleCardMinor