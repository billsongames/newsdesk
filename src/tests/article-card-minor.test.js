import React from "react";
import { render, screen, fireEvent, getNodeText } from "@testing-library/react";
import ArticleCardMinor from "../Components/ArticleCardMinor/ArticleCardMinor";

describe("ArticleCardMinor", () => {

  
  xit("renders correctly", () => {
    const validProps = {
      articles: [
        {
          title: "test_title_text",
          description: "test description text",
          content: "test description text",
          url: "test url",
          image: "test_image.jpg",
          publishedAt: "1970-01-01T00:0:00Z",
          source: {
          name: "test source",
            url: "test source url"
          }
        }
      ]
    }  

    render(<ArticleCardMinor props={validProps.articles} />);

    const { asFragment } = render(<ArticleCardMinor props={validProps.articles} />);

      expect(asFragment()).toMatchSnapshot();

  })
})