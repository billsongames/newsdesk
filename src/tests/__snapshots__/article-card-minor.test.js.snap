// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCardMinor renders correctly 1`] = `
<DocumentFragment>
  <div
    class="article-card-minor"
  >
    <div
      class="article-card-minor__title"
    />
    <div
      class="article-card-minor__display"
    >
      <div
        class="article-card-minor__image"
      >
        <img
          class="image"
        />
      </div>
      <div
        class="article-card-minor__text"
      >
        <div
          class="article-card-minor__description"
        />
        <div
          class="article-card-minor__content"
        />
      </div>
    </div>
    <div
      class="article-card-minor__url"
    >
      <a
        target="blank"
      >
        Full article
      </a>
    </div>
    <div
      class="article-card-minor__source"
    >
      Source: 
    </div>
    <div
      class="article-card-minor__time"
    />
  </div>
</DocumentFragment>
`;
